src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/LanguageServices/RegexPatternDetector.cs

```diff
@@ -0,0 +1,353 @@
+                        var constructor = _semanticModel.GetSymbolInfo(invocationOrCreation, cancellationToken).GetAnySymbol();
+                        if (_regexType.Equals(constructor?.ContainingType))
+                        {
+                            // Argument to "new Regex".  Need to do deeper analysis
+                            return AnalyzeStringLiteral(
+                                stringLiteral, argumentNode, cancellationToken, out options);
+                        }
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        public RegexTree TryParseRegexPattern(SyntaxToken token, CancellationToken cancellationToken)
+        {
+            if (!this.IsRegexPattern(token, cancellationToken, out var options))
+            {
+                return null;
+            }
+
+            var chars = _language.VirtualCharService.TryConvertToVirtualChars(token);
+            if (chars.IsDefaultOrEmpty)
+            {
+                return null;
+            }
+
+            return RegexParser.TryParse(chars, options);
+        }
+
+        private bool AnalyzeStringLiteral(
+            SyntaxToken stringLiteral, SyntaxNode argumentNode, 
+            CancellationToken cancellationToken, out RegexOptions options)
+        {
+            options = default;
+
+            var parameter = _language.SemanticFacts.FindParameterForArgument(_semanticModel, argumentNode, cancellationToken);
+            if (parameter?.Name != _patternName)
+            {
+                return false;
+            }
+
+            options = GetRegexOptions(argumentNode, cancellationToken);
+            return true;
+        }
+
+        private RegexOptions GetRegexOptions(SyntaxNode argumentNode, CancellationToken cancellationToken)
+        {
+            var syntaxFacts = _language.SyntaxFacts;
+            var argumentList = argumentNode.Parent;
+            var arguments = syntaxFacts.GetArgumentsOfArgumentList(argumentList);
+            foreach (var siblingArg in arguments)
+            {
+                if (siblingArg != argumentNode)
+                {
+                    var expr = syntaxFacts.GetExpressionOfArgument(siblingArg);
+                    if (expr != null)
+                    {
+                        var exprType = _semanticModel.GetTypeInfo(expr, cancellationToken);
+                        if (exprType.Type?.Name == nameof(RegexOptions))
+                        {
+                            var constVal = _semanticModel.GetConstantValue(expr, cancellationToken);
+                            if (constVal.HasValue)
+                            {
+                                return (RegexOptions)(int)constVal.Value;
+                            }
+                        }
+                    }
+                }
+            }
+
+            return RegexOptions.None;
+        }
+
+        private string GetNameOfType(SyntaxNode typeNode, ISyntaxFactsService syntaxFacts)
```

Comment:
>GetNameOfType [](start = 23, length = 13)

Would this method (and the following one) belong in `ISyntaxFactsService`?

Reply:
In general: no.

SyntaxFacts are not for general algorithms.  SyntaxFacts is really just to duplicate the existing syntax API in an agnostic way.  i.e. when both languages have effectively the same Syntax type with the same shape, then we can use SyntaxFacts to expose the same property-access operation on it. 

So, for example, both languages have InvocationExpression.  An Expression and an ArgList.  So SyntaxFacts gives you a way to say "give me the equivalent of InvocationExpression.Expression".

But the syntax API itself doesn't have something like type.GetName().  For this we use extensions.  And we use extensions heavily in the IDE layer.  But the point that we do that is when multiple features need to share the code.

Reply:
Note: we are not 100% consistent there in SyntaxFacts.  I would like to address that (given infinite time) :)